### PR TITLE
[3d] Terrain shading

### PR DESCRIPTION
**Pull request by wonder-sk**
This adds new options for user to choose how the terrain should be rendered:
- shading disabled - color of terrain is determined only from map texture
- shading enabled - color of terrain is determined using Phong's shading model,
  taking into account map texture, terrain normal vector, scene light(s) and
  terrain material's ambient+specular colors and shininess

(configuration of lights is not there yet - now we always have a single point
light with white color at the middle of the scene at 1km above the zero elevation)

Without shading:

terrain-without-shading

With shading (shininess=0):

terrain-with-shading-not-shiny

With shading (shininess=5):

terrain-with-shading

New configuration options in 3D view settings:
image

### CODE REVIEW

File: src/3d/qgs3dmapsettings.h [hunk 1]
```diff
@@ -200,6 +201,35 @@ class _3D_EXPORT Qgs3DMapSettings : public QObject
     //! Returns terrain generator. It takes care of producing terrain tiles from the input data.
     QgsTerrainGenerator *terrainGenerator() const { return mTerrainGenerator.get(); }
 
+    /**
+     * Sets whether terrain shading is enabled.
+     * \sa isTerrainShadingEnabled()
```
Comment (nyalldawson) [inline src/3d/qgs3dmapsettings.h]: I don't think sipify can handle "sa" - can you change this to "\see" instead?

**Review by nyalldawson — COMMENTED** [review]
Sweet!

**Comment by nirvn** [comment]
@wonder-sk , great feature. It might require to disable it for large scenes, it's not really functional:
image

File: src/3d/qgs3dmapsettings.h (hunk 1 above)
Comment (wonder-sk) [reply src/3d/qgs3dmapsettings.h]: updated in bcf8b0d

**Comment by wonder-sk** [comment]
As discussed with @nirvn in a private chat, configuration of lights is coming shortly...

**Comment by kannes** [comment]
Ambient occlusion might be a nice (fake) lighting system as well!

**Pull request merged**